core/io/image.cpp

```diff
@@ -1434,12 +1434,11 @@ int Image::_get_dst_image_size(int p_width, int p_height, Format p_format, int &
 		}
 
 		// Set mipmap size.
-		// It might be necessary to put this after the minimum mipmap size check because of the possible occurrence of "1 >> 1".
 		if (r_mm_width) {
-			*r_mm_width = bw >> 1;
+			*r_mm_width = w;
```

Comment:
Does this change fix a different bug or is it related to the other fixes in this PR? The connection is not obvious to me (being only somewhat familiar with the texture import code)

Reply:
I was having unexplained corruption with some texture dimensions and traced it to these returned width/height values (returned by `Image::get_image_mipmap_offset_and_dimensions`).

For example, on the 12x12 texture (block=4), I was seeing this. First loop iteration
```
{
  bw=12, bh=12
  s=144
  w = 6, h = 6
  *r_mm_width = (12>>1) = 6
  *r_mm_height = (12>>1) = 6
}
```
Second loop iteration
```
{
  bw=8, bh=8
  s=64
  w = 3, h = 3
  *r_mm_width = (8>>1) = 4 <-- wrong!
  *r_mm_height = (8>>1) = 4 <-- wrong!
}
```
Third loop iteration
```
{
  bw=4, bh=4
  s=16
  w = 1, h = 1
  *r_mm_width = (4>>1) = 2 <-- wrong!
  *r_mm_height = (4>>1) = 2 <-- wrong!
}
```

Note that the behavior of this function was altered slightly by 81ed9fa31bc63823235f64c88d441a8e530ab039
However, It was also wrong before. It's possible either that Godot always had this bug with 2x2 and 1x1 mipmaps reading out-of-bounds, or didn't cause an issue before because the old `etc` compression library ignored the out-of-bounds data.

I can't see any reason for the return values to use these "half-block" values. I would think if I want to know the dimensions of a mipmap level, I would want the actual dimensions, namely `w` and `h`. I can't find any callers of this code which would be adversely affected by this change.

In case you want to check my assertions, the only public functions using these `*r_mm_` return values are `Image::get_image_mipmap_size` and `Image::get_image_mipmap_offset_and_dimensions`.

I did some further digging into that logic, so a quick history breakdown. @reduz originally added this `*r_mm_width = bw;` statement in 0586e184490fd132f99acb1a67c788959cfdbade . This function was only used in one place, in what is now `servers/rendering/renderer_rd/renderer_storage_rd.cpp`. However, the function was only ever used with FORMAT_RGBA8 for the longest time, meaning block size is 1 and bw == w and bh == h. For this reason, I suspect reduz either never tested it with compressed formats, or had a different intended usecase in mind than what it has ultimately been used for.

Alright, I know I talk too much. I hope this answers your question about why this code change was necessary, why it was relevant to this PR, and why it evaded notice for so long. I'll also loop in @floppyhammer as they made the more recent commit here.

Reply:
What I did in https://github.com/godotengine/godot/commit/81ed9fa31bc63823235f64c88d441a8e530ab039 is just to make mipmap 0 correspond to the first mipmap instead of the original texture. The block size matching logic was there before that PR. As I haven't got a Vulkan-capable computer on hand right now, I will try to take a closer look at this on Monday.

Reply:
aha, and I think my version does the same thing, since I reference 'w' and 'h' after they are shifted by 1, compared with your code using bw shifted by 1. So I think in this sense, they both fulfill the requirement of having mipmap 0 correspond to the first mipmap.

My change should just be handling the non-multiple-of-four mipmaps a bit more carefully than how it was written before..

Reply:
Thanks for the detailed explanation. That all makes sense to me!

I see that ``Image::get_image_mipmap_offset_and_dimensions`` is also used here https://github.com/godotengine/godot/blob/c99c5a1caeafab161daebdde21ccd83ecdc897d5/scene/resources/texture.cpp#L400 But this code appears to only be used when loading VRAM compressed textures, so it makes sense that the logic mirrors the logic when compressing the texture in the first place.